src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -576,23 +589,46 @@ public ITableEntriesSnapshot GetSnapshot(int versionNumber)
 
             void IDisposable.Dispose()
             {
-                this.Presenter.AssertIsForeground();
+            }
 
-                // VS is letting go of us.  i.e. because a new FAR call is happening, or because
-                // of some other event (like the solution being closed).  Remove us from the set
-                // of sources for the window so that the existing data is cleared out.
-                Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 1);
-                Debug.Assert(_findReferencesWindow.Manager.Sources[0] == this);
+            #endregion
 
-                _findReferencesWindow.Manager.RemoveSource(this);
+            private void RemoveSink(ITableDataSink sink)
+            {
+                ImmutableInterlocked.Update(ref _tableDataSinks, (sinks, arg) => sinks.Remove(arg), sink);
 
-                CancelSearch();
+                if (_tableDataSinks.IsEmpty)
+                {
+                    // TODO: Why is this a safe assumption? Where is the contract that these subscriptions are only disposed on the main thread?
+                    //       Even if VS always does it, other folks can add their own sinks, and if the requirement isn't documented,
+                    //       others may dispose on another thread.
+                    Presenter.AssertIsForeground();
```

Comment:
This method is the one I'm least confident in. Which of these operations should happen when *all* sinks are removed vs. perhaps the removal of just one (specially recognized?) sink?

Also of course is my TODO comment about why this is assumed to be on the UI thread, and whether we can remove that requirement since I don't think that's a guarantee or requirement on the interface.

Reply:
this was communicated by the editor team as to how this will behave.  ANd our assert checks this (inlcuding in Release builds iirc) to verify this so that if that contract is violated we can inform them immediately.

Reply:
I am happy to accept that, but the docs on the interface should be updated to indicate that the interface must only be called from the UI thread.